Add optional structured pruning for YOLOv8 detection models

syedhamzamohiuddin opened the pull request:
### 🚀 Pruning Functionality for YOLO Detection Models

This PR introduces an optional **structured pruning** workflow for YOLOv8 detection models.
It supports pruning via either a global ratio (applied uniformly across components) or a custom per-layer configuration defined in YAML.

**✨ Key Features**

- **Global pruning:** `prune_ratio=0.1` prunes 10% of channels across supported components.
- **Config-based pruning:** `prune_yaml=path/to/prune.yaml` for fine-grained, per-layer ratios.
- **Supported layers:** `Conv`, `Bottleneck`, `C2f`, `SPPF`, and `Detect` (both classification & regression towers), with `Concat` handled transparently.
- **Round-trip support:** pruned models can be saved, reloaded, retrained, and exported.
- **No new dependencies:** implemented entirely with native PyTorch tensor ops.

**🧪 Tests**

Comprehensive test suite added in `tests/test_prune.py`:

- **Unit tests:** low-level primitives (`Conv2d`, `BatchNorm2d`, grouped convs, bias handling, skip connections).
- **Block-level pruning:** `Conv`, `Bottleneck`, `C2f`, `SPPF`, and `Detect` across YOLOv8/YOLOv11 variants.
- **Model-level validation:** mask propagation, channel continuity, and graph integrity across the full detection pipeline.
- **Round-trip checks:** prune → save → load → train/infer → export (PyTorch + ONNX) and `is_pruned` metadata persistence.
- **YAML schema validation:** malformed keys, invalid ratio types/ranges, and detect-head format errors.
- **Utilities:** helper functions (e.g., `gcd_all`) and groupwise pruning edge cases.

**📂 Config Example**

A sample config (`sample_prune.yaml`) is provided under `ultralytics/cfg/pruning/`,
demonstrating per-layer pruning ratios aligned with the `yolov8.yaml` architecture.

**📊 Accuracy/Speed Note (COCO128 Example)**

For a simple reproducibility benchmark, pruning was applied using ratios from `sample_prune.yaml`.
Both baseline and pruned models were exported to ONNX for *stability* and *reproducibility*.

| Model  (onnx)              | Epochs | Size   | inference speed (ms/img) | P     | R     | mAP50 | mAP50-95 |
|---------------------|--------|--------|----------------|-------|-------|-------|----------|
| YOLOv8s (baseline)  | 20     | 22.6 MB | 17.06 ms            | 0.878 | 0.787 | 0.869 | 0.697    |
| Pruned (retrain)| 55     | 11.6 MB | 12.03 ms            | 0.747 |0.607 | 0.687 | 0.488    |

Notes:
- Pruning reduces model size by **(\~49%)** and inference latency **(\~30%)**.  
- Expected accuracy drop observed pre-retraining; retraining helps recovery.
- Full reproducible notebook: [Colab Link](https://colab.research.google.com/drive/1mKClx1yqj7Y-H1T9zjve7onu6o1mgkIB?usp=sharing) 

**📝 Notes / Limitations**

- `torch.nn.utils.prune.ln_structured` was initially evaluated but excluded due to lack of **groupwise pruning** support (needed for depthwise/grouped convolutions and `C2f` splits).
- A custom pruning pipeline was implemented using raw PyTorch ops for full control.
- Unsupported components are skipped gracefully with clear log messages.
- Pruning is **fully optional** and backward-compatible.
- Currently supported: **YOLOv8 detection models only**.
- - Other families (YOLOv5, YOLOv7–YOLOv12, segmentation, pose, classification, etc.) are not yet supported.
- - Future extensions can be added if there’s interest.

**🧩 Summary**

This PR establishes the foundation for structured pruning in Ultralytics YOLO,
with a complete end-to-end workflow: prune → retrain → export → deploy.
Future work can extend support to other model families and pruning criteria.

syedhamzamohiuddin committed:
Add prune_detection_model functionality, config, and tests

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant commented:
👋 Hello @syedhamzamohiuddin, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process. An Ultralytics engineer will also assist you here shortly 😊

- ✅ Define a Purpose: Clearly explain the purpose of your fix or feature in your PR description, and link to any relevant issues. Ensure your commit messages are clear, concise, and adhere to the project's conventions. You can reference issues under the repository’s [Issues section](https://github.com/ultralytics/ultralytics/issues).
- ✅ Synchronize with Source: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ Ensure CI Checks Pass: Verify all Ultralytics Continuous Integration (CI) checks are passing. See our CI help here: [Ultralytics CI Guide](https://docs.ultralytics.com/help/CI/).
- ✅ Update Documentation: Update the relevant documentation for any new or modified features. Please include usage details for the new pruning workflow (global ratio and YAML-based configs) and examples in our docs: [Ultralytics Documentation](https://docs.ultralytics.com/).
- ✅ Add Tests: If applicable, include or update tests to cover your changes, and confirm that all tests are passing. Your added tests in `tests/test_prune.py` look great—please ensure they run under CI and cover edge cases (e.g., no-prune paths).
- ✅ Sign the CLA: Please ensure you have signed our Contributor License Agreement if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message. See: [Contributor License Agreement (CLA)](https://docs.ultralytics.com/help/CLA/).
- ✅ Minimize Changes: Limit your changes to the minimum necessary for your bug fix or feature addition. "It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is." — Bruce Lee

Additional notes for this PR:
- Please ensure the new module `ultralytics/utils/prune.py` is imported or referenced only where necessary, and that public APIs are clearly documented (e.g., `prune_detection_model`).
- Confirm that serialization markers like `ckpt["is_pruned"] = True` are consistently set and read across load/train flows, and reflected in docs for users.
- Since this feature currently targets YOLOv8 detection models, please add a short compatibility note in docs and inline comments referencing the provided sample config at `ultralytics/cfg/models/v8/sample_prune.yaml`.
- Verify that all new tests, including size reduction assertions, are stable across different environments and seeds.

For more guidance, please refer to our Contributing Guide: [Ultralytics Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! ✂️🚀

syedhamzamohiuddin committed:
Fix: use Optional[] for Python 3.8 compatibility in prune.py

syedhamzamohiuddin committed:
Add docstrings to prune tests for lint compliance

syedhamzamohiuddin committed:
Added return types to functions and recitfied docstring formatting

syedhamzamohiuddin committed:
appropriate placement of prune.yaml

syedhamzamohiuddin committed:
improvised tests for checking pruned model sizes

syedhamzamohiuddin committed:
added pruning support for pytorch 1.8. by adding torch.norm fallback

syedhamzamohiuddin commented:
Thanks for the feedback! Quick updates:

✅ No changes were made to default.yaml.

📊 Added a brief speed/accuracy note with results on COCO128 (plus Colab notebook link for reproducibility) in the PR description.

🧪 Current tests cover round-trip, YAML config, training, parameter reduction, and zero-prune sanity checks. I’m planning to extend with component-level checks (Conv, C2f, SPPF, Detect) for stricter unit coverage.

📄 For documentation, could you confirm the expected format? I’m thinking of adding a short .md page with a minimal usage example, current limitations, and tradeoffs.

❓ One clarification: the “tests guidance” link seems focused on model training/evaluation, not unit tests per se. Could you confirm what level of unit test coverage you’d like here?

hamza-swift commented:
Hi Glenn, just a quick update—I'm occupied with work commitments but still making progress here. Currently finalizing unit tests for pruning (component-level mask propagation, error paths, etc.) and will move on to higher-level modules next. Thanks for your patience!

syedhamzamohiuddin committed:
Added extesnsive testing and 'remove' type group pruning

syedhamzamohiuddin committed:
Added extesnsive testing and 'remove' type group pruning

syedhamzamohiuddin committed:
Added extesnsive testing and 'remove' type group pruning

syedhamzamohiuddin committed:
fix: use tensor-safe operations and modern type hints

- Use bitwise NOT (~) instead of comparison to False for tensor checks
- Add 'from __future__ import annotations' for Python 3.8 compatibility
- Replace Tuple with tuple in type hints to match Ultralytics style

syedhamzamohiuddin committed:
fix: added gcd wrapper for compatibility with python 3.8

syedhamzamohiuddin committed:
c

syedhamzamohiuddin committed:
undoing deletions

glenn-jocher commented:
Thanks—those failures are coming from `torch.onnx` importing `onnxscript`; the runner installs export deps so your local run likely skipped those tests. Please add `onnxscript` to our export deps (pyproject `[optional-dependencies].export`) or install it in CI, then re-run checks. To reproduce CI locally:

```bash
pip install -e ".[dev,export]" "onnxscript>=0.1.0"
pytest tests/test_exports.py::test_export_onnx -q
```

For CI setup details see the CI/Testing workflow guide, and for context on the `onnxscript` requirement see this PyTorch discussion.

syedhamzamohiuddin committed:
Add onnxscript to export dependencies for ONNX export tests

syedhamzamohiuddin committed:
Added missing return type annotations

syedhamzamohiuddin committed:
Modified sample_prune.yaml

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

syedhamzamohiuddin committed:
Merge branch 'main' into feature/prune-functionality

syedhamzamohiuddin commented:
Thanks --branch synced with main and ready for review

glenn-jocher commented:
Thanks—confirmed the branch is up to date; we’ll proceed with review and update here—please keep CI green and avoid new commits unless requested.
